master/buildbot/reporters/github.py

```diff
@@ -0,0 +1,158 @@
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
+"""Push events to github
+
+."""
+
+from datetime import datetime
+
+from buildbot import config
+from buildbot.util import service
+from buildbot.reporters import utils
+from buildbot.process.properties import Properties
+from buildbot.process.properties import Interpolate
+from buildbot.process.results import EXCEPTION
+from buildbot.process.results import FAILURE
+from buildbot.process.results import RETRY
+from buildbot.process.results import Results
+from buildbot.process.results import SUCCESS
+from buildbot.process.results import WARNINGS
+
+from twisted.internet import defer
+from twisted.internet import reactor
+from twisted.internet.protocol import ProcessProtocol
+from twisted.python import log
+
+try:
+    from txgithub.api import GithubApi as GitHubAPI
+except ImportError:
+    GitHubAPI = None
+
+class GitHubStatusPush(service.BuildbotService):
+
+    name = "GitHubStatusPush"
+
+    def reconfigService(self, token,
+            startDescription=None, endDescription=None,
+            context=None, baseURL=None, builders=None):
+        self.builders = builders
+
+        """
+        Token for GitHub API.
+        """
+        if not GitHubAPI:
+            config.error('GitHubStatus requires txgithub package installed')
+
+        self.context = context or Interpolate('buildbot/%(prop:buildername)s')
+        self.startDescription = startDescription or 'Build started.'
+        self.endDescription = endDescription or 'Build done.'
+
+        self.github = GitHubAPI(oauth2_token=token, baseURL=baseURL)
+
+        self.handledBuildIds = set()
+
+    @defer.inlineCallbacks
+    def startService(self):
+        yield service.BuildbotService.startService(self)
+        startConsuming = self.master.mq.startConsuming
+
+        self._buildUpdatedConsumer = yield startConsuming(
+            self.sendGitHubStatus,
+            ('builds', None, 'update'))
+
+        self._buildCompleteConsumer = yield startConsuming(
+            self.sendGitHubStatus,
+            ('builds', None, 'finished'))
+
+    def stopService(self):
+        self._buildUpdatedConsumer.stopConsuming()
+        self._buildCompleteConsumer.stopConsuming()
+
+    @defer.inlineCallbacks
+    def sendGitHubStatus(self, key, build):
+
+        if not build['complete'] and build['buildid'] in self.handledBuildIds:
+            return
+        else:
+            self.handledBuildIds.add(build['buildid'])
+
+        buildrequest = yield self.master.data.get(("buildrequests", build['buildrequestid']))
+        buildset = yield self.master.data.get(("buildsets", buildrequest['buildsetid']))
+        yield utils.getDetailsForBuilds(self.master, buildset, [build])
+        build['url'] = utils.getURLForBuild(self.master, build['builder']['builderid'], build['number'])
+
+        status = {}
+
+        props_dict = yield self.master.db.builds.getBuildProperties(build['buildid'])
```

Comment:
@sithhell  In this same block of code, needs to skip builds that haven't come from a webhook, else the whole thing crashes when parsing a non-existent sourcestamp. At my end I just skip if the scheduler is 'force' or one of the periodic schedulers and that seems to do the trick though I'm not sure if that's the best solution.